src/measure_theory/covering/vitali_family.lean

```diff
@@ -0,0 +1,227 @@
+subfamily `f x` of `sets_at x` containing sets of arbitrarily small diameter. Then one can extract
+a disjoint subfamily covering almost all `s`.
+
+Vitali families are provided by covering theorems such as the Besicovitch covering theorem or the
+Vitali covering theorem. They make it possible to formulate general versions of theorems on
+differentiations of measure that apply in both contexts.
+-/
+@[nolint has_inhabited_instance]
+structure vitali_family {m : measurable_space α} (μ : measure α) :=
+(sets_at : Π (x : α), set (set α))
+(measurable_set' : ∀ (x : α), ∀ (a : set α), a ∈ sets_at x → measurable_set a)
+(nonempty_interior : ∀ (x : α), ∀ (y : set α), y ∈ sets_at x → (interior y).nonempty)
+(nontrivial : ∀ (x : α) (ε > (0 : ℝ)), ∃ y ∈ sets_at x, y ⊆ closed_ball x ε)
+(covering : ∀ (s : set α) (f : Π (x : α), set (set α)), (∀ x ∈ s, f x ⊆ sets_at x) →
+  (∀ (x ∈ s) (ε > (0 : ℝ)), ∃ a ∈ f x, a ⊆ closed_ball x ε) →
+  ∃ (t : set α) (u : α → set α), t ⊆ s ∧ t.pairwise (disjoint on u) ∧ (∀ x ∈ t, u x ∈ f x)
+  ∧ μ (s \ ⋃ x ∈ t, u x) = 0)
+
+namespace vitali_family
+
+variables {m0 : measurable_space α} {μ : measure α}
+include μ
+
+/-- A Vitali family for a measure `μ` is also a Vitali family for any measure absolutely continuous
+with respect to `μ`. -/
+def mono (v : vitali_family μ) (ν : measure α) (hν : ν ≪ μ) :
+  vitali_family ν :=
+{ sets_at := v.sets_at,
+  measurable_set' := v.measurable_set',
+  nonempty_interior := v.nonempty_interior,
+  nontrivial := v.nontrivial,
+  covering := λ s f h h', begin
+    rcases v.covering s f h h' with ⟨t, u, ts, u_disj, uf, μu⟩,
+    exact ⟨t, u, ts, u_disj, uf, hν μu⟩
+  end }
+
+/-- Given a Vitali family `v` for a measure `μ`, a family `f` is a fine subfamily on a set `s` if
+every point `x` in `s` belongs to arbitrarily small sets in `v.sets_at x ∩ f x`. This is precisely
+the subfamilies for which the Vitali family definition ensures that one can extract a disjoint
+covering of almost all `s`. -/
+def fine_subfamily_on (v : vitali_family μ) (f : α → set (set α)) (s : set α) : Prop :=
+∀ x ∈ s, ∀ (ε > 0), ∃ a ∈ v.sets_at x ∩ f x, a ⊆ closed_ball x ε
+
+namespace fine_subfamily_on
+
+variables {v : vitali_family μ} {f : α → set (set α)} {s : set α} (h : v.fine_subfamily_on f s)
+include h
+
+theorem exists_disjoint_covering_ae :
+  ∃ (t : set α) (u : α → set α), t ⊆ s ∧ t.pairwise (disjoint on u) ∧
+    (∀ x ∈ t, u x ∈ v.sets_at x ∩ f x) ∧ μ (s \ ⋃ x ∈ t, u x) = 0 :=
+v.covering s (λ x, v.sets_at x ∩ f x) (λ x hx, inter_subset_left _ _) h
+
+/-- Given `h : v.fine_subfamily_on f s`, then `h.t` is a subset of `s` parametrizing a disjoint
+covering of almost every `s`. -/
+protected def t : set α :=
+h.exists_disjoint_covering_ae.some
+
+/-- Given `h : v.fine_subfamily_on f s`, then `h.u x` is a set in the family, for `x ∈ h.t`, such
+that these sets form a disjoint covering of almost every `s`. -/
+protected def u : α → set α :=
+h.exists_disjoint_covering_ae.some_spec.some
```

Comment:
Are these part of the API, or implementation details? If they are part of the API, I think they should have more descriptive names. If they are implementation details, please mention this in the docstring.